woocommerce/woocommerce-android
# Search in cache first on ProductSelection screen

`Do not merge label` - remove after [the FluxC PR](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2662) is merged and the reference is updated in build.gradle. 

This is still a draft but I'd like to know your opinion about the approach. I'll turn it into a proper PR when this approach is agreed on or close the PR if we opt for a different approach.

### Description


This PR adds support for local search to the ProductSelectorScreen. This screen is currently being used only in the Coupons flow, however, we plan to start using it in the Order Creation flow - the main aim of this PR is to reduce the time it takes to create an order.

The solution isn't perfect, but imo it gives us 90% of the functionality with minimal effort. 

`Caveat #1`
Imagine the following scenario:
1. User opens the search and enters a query
2. The app shows 25 products from the cache
3. The app initiates a request to fetch 25 search results from the server
4. The user scrolls to the bottom of the list
5. The app WILL NOT load more data from cache until request from step 3 completes successfully. 

This is a limitation of this solution, but I personally believe it's safe to assume that in most cases users who use search will enter enough letters so there are less than 25 results shown. Moreover, if the request completes, more data are loaded from the database. So the impact of this caveat is imo minimal. (implementing a proper solution isn't trivial)

`Caveat 2`
Products which are fetched after the products from cache are already shown are always appended to the bottom of the list - the products don't have any specific order. However, this feels better than changing the order of the products "under users' hands". Wdyt?

`Caveat 3`
Page 1 from the cache might not correspond to Page 1 from the server, but similar as in Caveat 1, I think even without it, it's an improvement and I assume most users will enter enough characters anyway. 

This PR also improves the current implementation of the search screen as it cancels previous requests when the user updates the search query -> previously, the app would wait for the previous request to complete before it initiated a new request.

This PR doesn't add any unit tests since the whole feature unfortunately isn't tested. I think it'd be nice to add tests for it, but that's out of the scope of this PR.

### Testing instructions


(You might need to hardcode `isCouponsEnabled=true` in `MoreMenuViewModel` in order to enable the feature)
1. Open More Menu
2. Tap on Coupons
3. Select a coupon
4. Tap on the overflow menu
5. Select Edit coupon
8. Tap on "Edit Products"
9. Test the search - you might want to try adding `delay(5000)` or changing the size of the page or generating a bunch of products on your site using [a plugin](https://wordpress.org/plugins/woocommerce-product-generator/). Make sure loading indicator is shown whenever a request is in progress - either during loadMore or when the search query changes.

### Images/gif
